### PR TITLE
Fixed-wing autotune

### PR DESCRIPTION
**Pull request by digitalentity**
AUTOTUNE will attempt to tune roll and pitch P, I and FF gains on a fixed-wing airplane.

Autotune will monitor behavior of the airplane when you fly it and adjust P, I and FF gains to reach optimal performance.

How to use:

Fly in ANGLE mode. Enable AUTOTUNE mode. Do hard maneuvers on each axis separately. For roll - bank hard left/hard right. For pitch - fast climb, steep dive. Initially you probably will notice very soft response - make sure your flying field is big enough for slow turns.

The more maneuvers you will do - the better results AUTOTUNE will be able to reach.

AUTOTUNE will adjust gains constantly but it will take a snapshot of current gains every 5 seconds. When you disable AUTOTUNE gains from last snapshot will be restored. If you turn AUTOTUNE on and off before 5 seconds elapse - PIFF gains won't be changed.

Currently AUTOTUNE don't save gains to EEPROM - you have to save manually, using a stick combo.

### CODE REVIEW

**Comment by Linjieqiang** [comment]
Is this branch OK to test?I am willing to have a try.

**Comment by digitalentity** [comment]
@Linjieqiang @oleost this should be ready to be tested. Use caution - as usual with highly experimental code. Be prepared to switch to PASSTHROUGH at all times :smile:

**Comment by brat002** [comment]
Maybe it will be more safe if PIDS restored back to initial state when out from autotune mode.

**Comment by digitalentity** [comment]
@brat002 how would you figure out the autotuned values then?

**Comment by Linjieqiang** [comment]
@digitalentity OK,I will test it later.So cool to see it.

**Comment by brat002** [comment]
@digitalentity it may be useful for fallback to safe default values. Not for previous, which can be also unsafe.

**Comment by digitalentity** [comment]
@brat002 you can't land in AUTOTUNE mode - near stall speeds will force autotune to ramp up the gains. You need a way to keep tuned gains AND disable AUTOTUNE in the air.

**Comment by digitalentity** [comment]
For safety you have PASSTHROUGH mode

**Comment by digitalentity** [comment]
Tested this today - it works. On my 80cm flying wing It brought PIDs from initial 20 on FF to 36/35 on pitch/roll FF which gave considerably better flight performance.

I flew about 1 minute with autotune enabled, landed w/o autotune and issued a "Save" stick command to store the new parameters.

**Comment by sppnk** [comment]
Wow, I will try to compile an hex this week and test myself !! Is blackbox available today for F1's?

**Comment by digitalentity** [comment]
Documentation is still lacking. What board are you using? I'll make a hex
for you.

**Comment by sppnk** [comment]
I could test in a cc3d_ppm1

**Comment by digitalentity** [comment]
@sppnk here you are
[inav_1.6.2_CC3D_PPM1_autotune.hex.gz](https://github.com/iNavFlight/inav/files/875278/inav_1.6.2_CC3D_PPM1_autotune.hex.gz)

File: src/main/flight/pid_autotune.c [hunk 1]
```diff
@@ -122,16 +123,31 @@ void autotuneUpdateState(void)
...
     }
 }
+
+static void blackboxLogAutotuneEvent(adjustmentFunction_e adjustmentFunction, int32_t newValue)
+{
+#ifndef BLACKBOX
+    UNUSED(adjustmentFunction);
+    UNUSED(newValue);
+#else
+    if (feature(FEATURE_BLACKBOX)) {
```
Comment (Linjieqiang) [inline src/main/flight/pid_autotune.c:152]: Maybe the head should add "#include "config/feature.h"" to avoid the warning info.

**Review by Linjieqiang — COMMENTED** [review]
I will fly my Z84 to check this cool feature tomorrow and report the test result.

File: src/main/flight/pid_autotune.c (hunk 1 above)
Comment (digitalentity) [reply src/main/flight/pid_autotune.c:152]: Good catch, thanks

**Pull request merged**